starboard/common/byte_swap.h

```diff
@@ -12,58 +12,67 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Module Overview: Starboard Byte Swap module C++ convenience layer
+// Module Overview: Starboard Byte Swap macros
 //
-// Implements convenient templated wrappers around the core Starboard Byte Swap
-// module. These functions are used to deal with endianness when performing I/O.
+// Specifies macros for swapping byte order. These functions are used to
+// deal with endianness when performing I/O.
 
 #ifndef STARBOARD_COMMON_BYTE_SWAP_H_
 #define STARBOARD_COMMON_BYTE_SWAP_H_
 
-#include <algorithm>
+#if SB_API_VERSION < 16
 
 #include "starboard/byte_swap.h"
-#include "starboard/types.h"
 
-namespace starboard {
+#else  // SB_API_VERSION < 16
 
-template <typename T>
-T SbByteSwap(T value) {
-  std::reverse(reinterpret_cast<uint8_t*>(&value),
-               reinterpret_cast<uint8_t*>(&value + 1));
-  return value;
-}
+#include "starboard/configuration.h"
+#include "starboard/types.h"
 
-template <>
-inline int16_t SbByteSwap(int16_t value) {
-  return SbByteSwapS16(value);
-}
+#ifdef __cplusplus
+extern "C" {
+#endif
 
-template <>
-inline uint16_t SbByteSwap(uint16_t value) {
-  return SbByteSwapU16(value);
+static __inline uint16_t __sb_bswap_16(uint16_t __x) {
+  return ((uint16_t)(__x << 8)) | __x >> 8;
```

Comment:
Consider using compiler intrinsics here. `#ifdef MSVC_VER _byteswap_ushort/_byteswap_ulong/_byteswap_uint64 #else __builtin_bswap16/__builtin_bswap32/__builtin_bswap64` should compile on every compiler.

There's a theoretical benefit of getting directly emitted native byte swap instructions on all architectures, but tbh the optimizer would likely do that anyway here.

Reply:
Let's add a TODO comment and file a ticket for the builtins.
This cleanup is great! 

Most probably we are not regressing the performance as we were already doing function call before and now it all inlined, but the fact that the compilers have separate builtin functions for those operations suggests that that maybe they can't always do the job optimally.